File: src/main/node/net/ecma/data/ViewPositionOrientationPrototype.mjs
'use strict';
import fs from 'fs';
import { X3D } from './x3d.mjs';
import { SFString } from './x3d.mjs';
import { SFNode } from './x3d.mjs';
import { head } from './x3d.mjs';
import { MFNode } from './x3d.mjs';
import { meta } from './x3d.mjs';
import { Scene } from './x3d.mjs';
import { ProtoDeclare } from './x3d.mjs';
import { ProtoInterface } from './x3d.mjs';
import { field } from './x3d.mjs';
import { ProtoBody } from './x3d.mjs';
import { ProximitySensor } from './x3d.mjs';
import { SFVec3f } from './x3d.mjs';
import { IS } from './x3d.mjs';
import { connect } from './x3d.mjs';
import { Script } from './x3d.mjs';
import { ROUTE } from './x3d.mjs';
import { Viewpoint } from './x3d.mjs';
import { Anchor } from './x3d.mjs';
import { MFString } from './x3d.mjs';
import { Shape } from './x3d.mjs';
import { Text } from './x3d.mjs';
import { FontStyle } from './x3d.mjs';
import { SFFloat } from './x3d.mjs';
import { Appearance } from './x3d.mjs';
import { Material } from './x3d.mjs';
import { SFColor } from './x3d.mjs';
import { Box } from './x3d.mjs';
var X3D0 =  new X3D({

      profile : new SFString("Immersive"),
      version : new SFString("3.0"),
      head : new SFNode(
        new head({
          meta : new MFNode([
            new meta({
              name : new SFString("title"),
              content : new SFString("ViewPositionOrientationPrototype.x3d")}),

            new meta({
              name : new SFString("creator"),
              content : new SFString("Leonard Daly and Don Brutzman")}),

            new meta({
              name : new SFString("translator"),
              content : new SFString("Don Brutzman")}),

            new meta({
              name : new SFString("created"),
              content : new SFString("1 November 2000")}),

            new meta({
              name : new SFString("modified"),
              content : new SFString("27 May 2017")}),

            new meta({
              name : new SFString("description"),
              content : new SFString("ViewPositionOrientation prototype provides local position and orientation as user navigates, with optional console output")}),

            new meta({
              name : new SFString("info"),
              content : new SFString("Note fix: metadata is no longer an allowed ProtoDeclare field name, since ProtoInstance already includes a metadata field")}),

            new meta({
              name : new SFString("reference"),
              content : new SFString("ViewPositionOrientationExample.x3d")}),

            new meta({
              name : new SFString("reference"),
              content : new SFString("http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter14Prototypes/WhereAmIExample.x3d")}),

            new meta({
              name : new SFString("reference"),
              content : new SFString("http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter14Prototypes/WhereAmIPrototype.x3d")}),

            new meta({
              name : new SFString("reference"),
              content : new SFString("http://www.realism.com/Web3D/Examples#WhereAmI")}),

            new meta({
              name : new SFString("reference"),
              content : new SFString("http://www.realism.com/vrml/Example/WhereAmI/WhereAmI_Proto.wrl")}),

            new meta({
              name : new SFString("identifier"),
              content : new SFString("https://savage.nps.edu/Savage/Tools/Authoring/ViewPositionOrientationPrototype.x3d")}),

            new meta({
              name : new SFString("generator"),
              content : new SFString("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit")}),

            new meta({
              name : new SFString("license"),
              content : new SFString("../../license.html")})])})),
      Scene : new SFNode(
        new Scene({
          /*Thanks to Leonard Daly for the original design of this prototype*/
          children : new MFNode([
            new ProtoDeclare({
              name : new SFString("ViewPositionOrientation"),
              appinfo : new SFString("ViewPositionOrientation provides provides console output of local position and orientation as user navigates"),
              ProtoInterface : new SFNode(
                new ProtoInterface({
                  field : new MFNode([
                    new field({
                      type : field.TYPE_SFBOOL,
                      name : new SFString("enabled"),
                      accessType : new SFString(field.ACCESSTYPE_INPUTOUTPUT),
                      appinfo : new SFString("Whether or not ViewPositionOrientation sends output to console"),
                      value : new SFString("true")}),

                    new field({
                      type : field.TYPE_SFBOOL,
                      name : new SFString("traceEnabled"),
                      accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                      appinfo : new SFString("Output internal trace messages for debugging this node, intended for developer use only"),
                      value : new SFString("true")}),

                    new field({
                      type : field.TYPE_SFBOOL,
                      name : new SFString("set_traceEnabled"),
                      accessType : new SFString(field.ACCESSTYPE_INPUTONLY),
                      appinfo : new SFString("Ability to turn output tracing on/off at runtime")}),

                    new field({
                      type : field.TYPE_SFVEC3F,
                      name : new SFString("position_changed"),
                      accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY),
                      appinfo : new SFString("Output local position")}),

                    new field({
                      type : field.TYPE_SFROTATION,
                      name : new SFString("orientation_changed"),
                      accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY),
                      appinfo : new SFString("Output local orientation")}),

                    new field({
                      type : field.TYPE_MFSTRING,
                      name : new SFString("outputViewpointString"),
                      accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY),
                      appinfo : new SFString("MFString value of new Viewpoint, suitable for use in string field of a Text node")})])})),
              ProtoBody : new SFNode(
                new ProtoBody({
                  children : new MFNode([
                    new ProximitySensor({
                      DEF : new SFString("WhereSensor"),
                      size : new SFVec3f([1000000000,1000000000,1000000000]),
                      IS : new SFNode(
                        new IS({
                          connect : new MFNode([
                            new connect({
                              nodeField : new SFString("enabled"),
                              protoField : new SFString("enabled")})])}))}),
                  /*it's a big old world out there! large box likely handles most use cases.*/

                    new Script({
                      DEF : new SFString("OutputPositionOrientation"),
                      field : new MFNode([
                        new field({
                          type : field.TYPE_SFBOOL,
                          name : new SFString("traceEnabled"),
                          accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY)}),

                        new field({
                          type : field.TYPE_SFBOOL,
                          name : new SFString("set_traceEnabled"),
                          accessType : new SFString(field.ACCESSTYPE_INPUTONLY)}),

                        new field({
                          type : field.TYPE_SFVEC3F,
                          name : new SFString("set_position"),
                          accessType : new SFString(field.ACCESSTYPE_INPUTONLY)}),

                        new field({
                          type : field.TYPE_SFROTATION,
                          name : new SFString("set_orientation"),
                          accessType : new SFString(field.ACCESSTYPE_INPUTONLY)}),

                        new field({
                          type : field.TYPE_SFVEC3F,
                          name : new SFString("position"),
                          accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                          value : new SFString("0 0 0")}),

                        new field({
                          type : field.TYPE_SFROTATION,
                          name : new SFString("orientation"),
                          accessType : new SFString(field.ACCESSTYPE_INITIALIZEONLY),
                          value : new SFString("0 1 0 0")}),

                        new field({
                          type : field.TYPE_SFVEC3F,
                          name : new SFString("position_changed"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY)}),

                        new field({
                          type : field.TYPE_SFROTATION,
                          name : new SFString("orientation_changed"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY)}),

                        new field({
                          type : field.TYPE_MFSTRING,
                          name : new SFString("outputViewpointString"),
                          accessType : new SFString(field.ACCESSTYPE_OUTPUTONLY)}),
                      IS : new SFNode(
                        new IS({
                          connect : new MFNode([
                            new connect({
                              nodeField : new SFString("traceEnabled"),
                              protoField : new SFString("traceEnabled")}),

                            new connect({
                              nodeField : new SFString("set_traceEnabled"),
                              protoField : new SFString("set_traceEnabled")}),

                            new connect({
                              nodeField : new SFString("position_changed"),
                              protoField : new SFString("position_changed")}),

                            new connect({
                              nodeField : new SFString("orientation_changed"),
                              protoField : new SFString("orientation_changed")}),

                            new connect({
                              nodeField : new SFString("outputViewpointString"),
                              protoField : new SFString("outputViewpointString")})])})]),
                      
ecmascript:eval (0

, function roundoff (value, digits) // for local use only
{
	var resolution = 1;
	for (i = 1; i <= digits; i++ )
	{
		resolution *= 10;
	}
	return Math.round (value*resolution) / resolution; // round to resolution
}
, function outputViewpoint ()
{
  var holdString = 
        '<Viewpoint position="' +
		roundoff (position.x, 1) + ' ' +
		roundoff (position.y, 1) + ' ' +
		roundoff (position.z, 1) +
        '" orientation="' +
		roundoff (orientation.x, 3) + ' ' +
		roundoff (orientation.y, 3) + ' ' +
		roundoff (orientation.z, 3) + ' ' +
		roundoff (orientation.angle, 4) + '"/>' ;
   tracePrint (holdString);
   outputViewpointString = new MFString (holdString);
}
, function set_position (value)
{
	position = value; // save persistent value
	position_changed = position; // output event
	outputViewpoint ();
}
, function set_orientation (value)
{
	orientation = value; // save persistent value
	orientation_changed = orientation; // output event
	outputViewpoint ();
}
, function set_traceEnabled (value)
{
	traceEnabled = value;
	alwaysPrint ('traceEnabled=' + traceEnabled);
}
, function alwaysPrint (text)
{
	Browser.print ('[ViewPositionOrientation] ' + text + '\n');
}
, function tracePrint (text)
{
	if (traceEnabled) alwaysPrint (text);
})}),

                    new ROUTE({
                      fromField : new SFString("position_changed"),
                      fromNode : new SFString("WhereSensor"),
                      toField : new SFString("set_position"),
                      toNode : new SFString("OutputPositionOrientation")}),

                    new ROUTE({
                      fromField : new SFString("orientation_changed"),
                      fromNode : new SFString("WhereSensor"),
                      toField : new SFString("set_orientation"),
                      toNode : new SFString("OutputPositionOrientation")})])}))}),
          /*====================*/

            new Viewpoint({
              description : new SFString("ViewPositionOrientation prototype"),
              position : new SFVec3f([0,0,14])}),

            new Anchor({
              description : new SFString("ViewPositionOrientation Example"),
              parameter : new MFString(["target=_blank"]),
              url : new MFString(["ViewPositionOrientationExample.x3d","https://savage.nps.edu/Savage/Tools/Authoring/ViewPositionOrientationExample.x3d","ViewPositionOrientationExample.wrl","https://savage.nps.edu/Savage/Tools/Authoring/ViewPositionOrientationExample.wrl"]),
              children : new MFNode([
                new Shape({
                  geometry : new SFNode(
                    new Text({
                      string : new MFString(["ViewPositionOrientationPrototype","is a prototype definition file","","Click this text to see","ViewPositionOrientationExample"]),
                      fontStyle : new SFNode(
                        new FontStyle({
                          justify : new MFString(["MIDDLE","MIDDLE"]),
                          size : new SFFloat(1.2)}))})),
                  appearance : new SFNode(
                    new Appearance({
                      material : new SFNode(
                        new Material({
                          diffuseColor : new SFColor([0.6,0.8,0.4])}))}))}),
              /*Selectable Text design pattern has transparent Box and TouchSensor description as a tooltip*/

                new Shape({
                  /*Author TODO: to adjust transparent Box as text-selection assist, set width and height to match size, then set transparency='1' to make invisible.*/
                  geometry : new SFNode(
                    new Box({
                      size : new SFVec3f([16,7,0.001])})),
                  appearance : new SFNode(
                    new Appearance({
                      material : new SFNode(
                        new Material({
                          transparency : new SFFloat(1)}))}))})])})])}))});
console.log(X3D0.toXMLNode());

File: src/main/python/net/x3djsonld/data/c/x3d-code/www.web3d.org/x3d/content/examples/Savage/Tools/Authoring/ViewPositionOrientationPrototype.py
###############################################
#
# Now available: developmental python x3d.py package on PyPi for import.
#   This approach greatly simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.

from x3d import *

###############################################

newModel=X3D(profile='Immersive',version='3.0',
  head=head(
    children=[
    meta(content='ViewPositionOrientationPrototype.x3d',name='title'),
    meta(content='Leonard Daly and Don Brutzman',name='creator'),
    meta(content='Don Brutzman',name='translator'),
    meta(content='1 November 2000',name='created'),
    meta(content='27 May 2017',name='modified'),
    meta(content='ViewPositionOrientation prototype provides local position and orientation as user navigates, with optional console output',name='description'),
    meta(content='Note fix: metadata is no longer an allowed ProtoDeclare field name, since ProtoInstance already includes a metadata field',name='info'),
    meta(content='ViewPositionOrientationExample.x3d',name='reference'),
    meta(content='http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter14Prototypes/WhereAmIExample.x3d',name='reference'),
    meta(content='http://X3dGraphics.com/examples/X3dForWebAuthors/Chapter14Prototypes/WhereAmIPrototype.x3d',name='reference'),
    meta(content='http://www.realism.com/Web3D/Examples#WhereAmI',name='reference'),
    meta(content='http://www.realism.com/vrml/Example/WhereAmI/WhereAmI_Proto.wrl',name='reference'),
    meta(content='https://savage.nps.edu/Savage/Tools/Authoring/ViewPositionOrientationPrototype.x3d',name='identifier'),
    meta(content='X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit',name='generator'),
    meta(content='../../license.html',name='license')]),
  Scene=Scene(
    #  Thanks to Leonard Daly for the original design of this prototype 
    children=[
    ProtoDeclare(appinfo='ViewPositionOrientation provides provides console output of local position and orientation as user navigates',name='ViewPositionOrientation',
      ProtoInterface=ProtoInterface(
        field=[
        field(accessType='inputOutput',appinfo='Whether or not ViewPositionOrientation sends output to console',name='enabled',type='SFBool',value=True),
        field(accessType='initializeOnly',appinfo='Output internal trace messages for debugging this node, intended for developer use only',name='traceEnabled',type='SFBool',value=True),
        field(accessType='inputOnly',appinfo='Ability to turn output tracing on/off at runtime',name='set_traceEnabled',type='SFBool'),
        field(accessType='outputOnly',appinfo='Output local position',name='position_changed',type='SFVec3f'),
        field(accessType='outputOnly',appinfo='Output local orientation',name='orientation_changed',type='SFRotation'),
        field(accessType='outputOnly',appinfo='MFString value of new Viewpoint, suitable for use in string field of a Text node',name='outputViewpointString',type='MFString')]),
      ProtoBody=ProtoBody(
        children=[
        ProximitySensor(DEF='WhereSensor',size=(1000000000,1000000000,1000000000),
          IS=IS(
            connect=[
            connect(nodeField='enabled',protoField='enabled')])),
        #  it's a big old world out there! large box likely handles most use cases. 
        Script(DEF='OutputPositionOrientation',
          field=[
          field(accessType='initializeOnly',name='traceEnabled',type='SFBool'),
          field(accessType='inputOnly',name='set_traceEnabled',type='SFBool'),
          field(accessType='inputOnly',name='set_position',type='SFVec3f'),
          field(accessType='inputOnly',name='set_orientation',type='SFRotation'),
          field(accessType='initializeOnly',name='position',type='SFVec3f',value=(0,0,0)),
          field(accessType='initializeOnly',name='orientation',type='SFRotation',value=(0,1,0,0)),
          field(accessType='outputOnly',name='position_changed',type='SFVec3f'),
          field(accessType='outputOnly',name='orientation_changed',type='SFRotation'),
          field(accessType='outputOnly',name='outputViewpointString',type='MFString')],
          IS=IS(
            connect=[
            connect(nodeField='traceEnabled',protoField='traceEnabled'),
            connect(nodeField='set_traceEnabled',protoField='set_traceEnabled'),
            connect(nodeField='position_changed',protoField='position_changed'),
            connect(nodeField='orientation_changed',protoField='orientation_changed'),
            connect(nodeField='outputViewpointString',protoField='outputViewpointString')])),
        ROUTE(fromField='position_changed',fromNode='WhereSensor',toField='set_position',toNode='OutputPositionOrientation'),
        ROUTE(fromField='orientation_changed',fromNode='WhereSensor',toField='set_orientation',toNode='OutputPositionOrientation')])),
    #  ==================== 
    Viewpoint(description='ViewPositionOrientation prototype',position=(0,0,14)),
    Anchor(description='ViewPositionOrientation Example',parameter=["target=_blank"],url=["ViewPositionOrientationExample.x3d","https://savage.nps.edu/Savage/Tools/Authoring/ViewPositionOrientationExample.x3d","ViewPositionOrientationExample.wrl","https://savage.nps.edu/Savage/Tools/Authoring/ViewPositionOrientationExample.wrl"],
      children=[
      Shape(
        geometry=Text(string=["ViewPositionOrientationPrototype","is a prototype definition file","","Click this text to see","ViewPositionOrientationExample"],
          fontStyle=FontStyle(justify=["MIDDLE","MIDDLE"],size=1.2)),
        appearance=Appearance(
          material=Material(diffuseColor=(0.6,0.8,0.4)))),
      #  Selectable Text design pattern has transparent Box and TouchSensor description as a tooltip 
      Shape(
        #  Author TODO: to adjust transparent Box as text-selection assist, set width and height to match size, then set transparency='1' to make invisible. 
        geometry=Box(size=(16,7,.001)),
        appearance=Appearance(
          material=Material(transparency=1)))])])
) # X3D model complete

###############################################
# Self-test diagnostics
###############################################

if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel))
print('check  newModel.XML() serialization...')
newModelXML = newModel.XML() # test export method XML() for exceptions
# print(newModelXML) # debug

print ("python x3d.py load successful for ViewPositionOrientationPrototype.py")
